From 8bd993cf28784c0f6107f7ec1cdc149ba8a39b20 Mon Sep 17 00:00:00 2001
From: Jun Zhang <jun.zhang@arm.com>
Date: Fri, 24 Nov 2023 11:39:12 +0800
Subject: [PATCH] rss/kronos: Parse GPT partition for AP secure flash

AP secure flash will also use the GPT partition to
store Metadata and FIP image, so parsing the partition
is needed for RSS to load AP BL2 from FIP in AP
secure flash later.

Upstream-Status: Pending
Signed-off-by: Jun Zhang <jun.zhang@arm.com>
---
 .../target/arm/rss/kronos/bl2/boot_hal_bl2.c  | 28 +++++++++----
 .../arm/rss/kronos/partition/partition.c      |  2 +
 platform/ext/target/arm/rss/kronos/platform.c | 41 ++++++++++++++-----
 platform/ext/target/arm/rss/kronos/platform.h |  5 ++-
 4 files changed, 55 insertions(+), 21 deletions(-)

diff --git a/platform/ext/target/arm/rss/kronos/bl2/boot_hal_bl2.c b/platform/ext/target/arm/rss/kronos/bl2/boot_hal_bl2.c
index 3d055082c..33d200082 100644
--- a/platform/ext/target/arm/rss/kronos/bl2/boot_hal_bl2.c
+++ b/platform/ext/target/arm/rss/kronos/bl2/boot_hal_bl2.c
@@ -36,6 +36,7 @@
 #endif /* CRYPTO_HW_ACCELERATOR */
 #include "platform.h"
 #include "partition.h"
+#include "fwu_agent.h"

 extern volatile bool scp_doorbell;
 extern ARM_DRIVER_FLASH FLASH_DEV_NAME;
@@ -59,8 +60,6 @@ static const char * const image_names [][2] = {
     {"scp_primary", "scp_secondary"},
 };

-#define BANK_0                  (0)
-
 static bool fill_rss_flash_map_with_data(uint8_t boot_index) {
     partition_entry_t *entry = NULL;
     uint8_t i = 0;
@@ -431,6 +430,7 @@ int32_t tower_nci_periph_init(void) {
 int32_t boot_platform_init(void)
 {
     int32_t result;
+    uint8_t boot_index;

     read_chip_id();

@@ -454,17 +454,21 @@ int32_t boot_platform_init(void)
         return 1;
     }

-    plat_io_storage_init();
-    partition_init(PLATFORM_GPT_IMAGE);
+    /* AP flash initial */
+    result = AP_FLASH_DEV_NAME.Initialize(NULL);
+    if (result != ARM_DRIVER_OK) {
+        return 1;
+    }

-    if (!fill_rss_flash_map_with_data(0)) {
-        BOOT_LOG_ERR("Filling flash map with signed images has failed!");
+    result = plat_io_storage_init(PLATFORM_GPT_IMAGE_RSS);
+    if (result != 0) {
         return 1;
     }

-    /* AP flash initial */
-    result = AP_FLASH_DEV_NAME.Initialize(NULL);
-    if (result != ARM_DRIVER_OK) {
+    partition_init(PLATFORM_GPT_IMAGE_RSS);
+    boot_index = bl2_get_boot_bank();
+    if (!fill_rss_flash_map_with_data(boot_index)) {
+        BOOT_LOG_ERR("Filling flash map with signed images has failed!");
         return 1;
     }

@@ -1044,6 +1048,7 @@ static int boot_platform_post_load_lcp(void)
 static int boot_platform_pre_load_ap_bl2(void)
 {
     enum atu_error_t atu_err;
+    int32_t result;

     BOOT_LOG_INF("BL2: AP BL2 pre load start");

@@ -1077,6 +1082,11 @@ static int boot_platform_pre_load_ap_bl2(void)
         return 1;
     }

+    result = plat_io_storage_init(PLATFORM_GPT_IMAGE_AP);
+    if (result != 0) {
+        return 1;
+    }
+    partition_init(PLATFORM_GPT_IMAGE_AP);
     BOOT_LOG_INF("BL2: AP BL2 pre load complete");

     return 0;
diff --git a/platform/ext/target/arm/rss/kronos/partition/partition.c b/platform/ext/target/arm/rss/kronos/partition/partition.c
index fe6f229cd..2b854e759 100644
--- a/platform/ext/target/arm/rss/kronos/partition/partition.c
+++ b/platform/ext/target/arm/rss/kronos/partition/partition.c
@@ -19,6 +19,8 @@
 #include "io_storage.h"
 #include "platform.h"
 #include "soft_crc.h"
+#include "host_base_address.h"
+#include "io_driver.h"

 static uint8_t mbr_sector[PLAT_PARTITION_BLOCK_SIZE];
 static partition_entry_list_t list;
diff --git a/platform/ext/target/arm/rss/kronos/platform.c b/platform/ext/target/arm/rss/kronos/platform.c
index 623811764..c650390b6 100644
--- a/platform/ext/target/arm/rss/kronos/platform.c
+++ b/platform/ext/target/arm/rss/kronos/platform.c
@@ -17,48 +17,69 @@
 typedef struct {
     uintptr_t dev_handle;
     uintptr_t image_spec;
+    uintptr_t flash_dev_con;
+    uintptr_t flash_dev_spec;
 } platform_image_source_t;

 extern ARM_DRIVER_FLASH FLASH_DEV_NAME;
+extern ARM_DRIVER_FLASH Driver_FLASH1;

 #define FLASH_SECTOR_SIZE 0x1000

-static io_dev_connector_t *flash_dev_con;
 static uint8_t local_block_flash[FLASH_SECTOR_SIZE];
-static io_flash_dev_spec_t flash_dev_spec = {
+static io_flash_dev_spec_t flash_dev_spec_rss = {
     .buffer = local_block_flash,
     .bufferlen = FLASH_SECTOR_SIZE,
     .flash_driver = &FLASH_DEV_NAME,
 };
-static io_block_spec_t flash_spec = {
+static io_block_spec_t flash_spec_rss = {
     .length = FLASH_TOTAL_SIZE
 };

+static io_flash_dev_spec_t flash_dev_spec_ap = {
+    .buffer = local_block_flash,
+    .bufferlen = FLASH_SECTOR_SIZE,
+    .flash_driver = &Driver_FLASH1,
+};
+
+static io_block_spec_t flash_spec_ap = {
+    .length = AP_FLASH_SIZE
+};
 static platform_image_source_t platform_image_source[] = {
-    [PLATFORM_GPT_IMAGE] = {
+    [PLATFORM_GPT_IMAGE_RSS] = {
+        .dev_handle = NULL,
+        .image_spec = &flash_spec_rss,
+        .flash_dev_con = NULL,
+        .flash_dev_spec = &flash_dev_spec_rss,
+    },
+
+    [PLATFORM_GPT_IMAGE_AP] = {
         .dev_handle = NULL,
-        .image_spec = &flash_spec,
+        .image_spec = &flash_spec_ap,
+        .flash_dev_con = NULL,
+        .flash_dev_spec = &flash_dev_spec_ap,
     }
 };

 /* Initialize io storage of the platform */
-int32_t plat_io_storage_init(void)
+int32_t plat_io_storage_init(uint8_t gpt_img_id)
 {
     int rc = -1;
     uintptr_t flash_dev_handle = NULL;
     uintptr_t flash_handle = NULL;

-    rc = register_io_dev_flash((const io_dev_connector_t **) &flash_dev_con);
+    rc = register_io_dev_flash((const io_dev_connector_t **) &platform_image_source[gpt_img_id].flash_dev_con);
     if (rc != 0) {
         return rc;
     }

-    rc = io_dev_open(flash_dev_con, (const uintptr_t)&flash_dev_spec, &flash_dev_handle);
+    rc = io_dev_open(platform_image_source[gpt_img_id].flash_dev_con,
+                     (const uintptr_t)platform_image_source[gpt_img_id].flash_dev_spec, &flash_dev_handle);
     if (rc != 0) {
         return rc;
     }

-    rc = io_open(flash_dev_handle, (const uintptr_t)&flash_spec, &flash_handle);
+    rc = io_open(flash_dev_handle, (const uintptr_t)platform_image_source[gpt_img_id].image_spec, &flash_handle);
     if (rc != 0) {
         return rc;
     }
@@ -68,7 +89,7 @@ int32_t plat_io_storage_init(void)
         return rc;
     }
     /* Update the platform image source that uses the flash with dev handles */
-    platform_image_source[PLATFORM_GPT_IMAGE].dev_handle = flash_dev_handle;
+    platform_image_source[gpt_img_id].dev_handle = flash_dev_handle;

     return rc;
 }
diff --git a/platform/ext/target/arm/rss/kronos/platform.h b/platform/ext/target/arm/rss/kronos/platform.h
index a88093ed4..ece472ee7 100644
--- a/platform/ext/target/arm/rss/kronos/platform.h
+++ b/platform/ext/target/arm/rss/kronos/platform.h
@@ -9,7 +9,8 @@
 #define __PLATFORM_H__

 typedef enum {
-    PLATFORM_GPT_IMAGE = 0,
+    PLATFORM_GPT_IMAGE_RSS = 0,
+    PLATFORM_GPT_IMAGE_AP,
     PLATFORM_IMAGE_COUNT,
 }platform_image_id_t;

@@ -19,7 +20,7 @@ typedef enum {
      ((uuid_t){{0xc3, 0x5d, 0xb5, 0xec}, {0xb7, 0x8a}, {0x84, 0x4a}, 0xab,  0x56, {0xeb, 0x0a, 0x99, 0x74, 0xdb, 0x42}})

 /* Initialize io storage of the platform */
-int32_t plat_io_storage_init(void);
+int32_t plat_io_storage_init(uint8_t gpt_img_id);

 /* Return an IO device handle and specification which can be used to access
  * an image. This has to be implemented for the GPT parser. */
--
2.25.1

